ktor-io/common/src/io/ktor/utils/io/ByteReadChannelOperations.kt

```diff
@@ -490,3 +491,173 @@ public fun ByteWriteChannel.rethrowCloseCauseIfNeeded() {
+ * @return The number of bytes read, not including the search string.
+ * @throws IOException If the limit is exceeded or the byteString is not found and ignoreMissing is false.
+ */
+public suspend fun ByteReadChannel.readUntil(
+    matchString: ByteString,
+    writeChannel: ByteWriteChannel,
+    limit: Long = Long.MAX_VALUE,
+    ignoreMissing: Boolean = false,
+): Long {
+    check(matchString.size > 0) {
+        "Empty match string not permitted for readUntil"
+    }
+    val partialMatchTable = buildPartialMatchTable(matchString)
+    var matchIndex = 0
+    val matchBuffer = ByteArray(matchString.size)
+    var rc = 0L
+
+    suspend fun appendPartialMatch() {
+        writeChannel.writeFully(matchBuffer, 0, matchIndex)
+        rc += matchIndex
+        matchIndex = 0
+    }
+
+    fun resetPartialMatch(byte: Byte) {
+        while (matchIndex > 0 && byte != matchString[matchIndex]) {
+            matchIndex = partialMatchTable[matchIndex - 1]
+        }
+    }
+
+    while (!isClosedForRead) {
+        val byte = readByte()
+
+        if (matchIndex > 0 && byte != matchString[matchIndex]) {
+            appendPartialMatch()
+            resetPartialMatch(byte)
+        }
+
+        if (byte == matchString[matchIndex]) {
+            matchBuffer[matchIndex] = byte
+            if (++matchIndex == matchString.size) {
+                return rc
+            }
+        } else {
+            writeChannel.writeByte(byte)
+            rc++
+        }
+
+        if (rc > limit) {
+            throw IOException("Limit of $limit bytes exceeded while scanning for \"${matchString.decodeToString()}\"")
+        }
+    }
+
+    if (ignoreMissing) {
+        appendPartialMatch()
+        writeChannel.flush()
+        return rc
+    }
+
+    throw IOException("Expected \"${matchString.toSingleLineString()}\" but encountered end of input")
+}
+
+/**
+ * Helper function to build the partial match table (also known as "longest prefix suffix" table)
+ */
+private fun buildPartialMatchTable(byteString: ByteString): IntArray {
+    val table = IntArray(byteString.size)
+    var j = 0
+
+    for (i in 1 until byteString.size) {
+        while (j > 0 && byteString[i] != byteString[j]) {
+            j = table[j - 1]
+        }
+        if (byteString[i] == byteString[j]) {
+            j++
+        }
+        table[i] = j
+    }
+
+    return table
+}
+
+/**
+ * Attempts to skip over a sequence of bytes in the channel that matches the provided [byteString].
+ *
+ * @param byteString The byte sequence to be matched and skipped over.
+ * @return [SkipResult.Ok] if the sequence matches and is successfully skipped,
+ *         [SkipResult.EndOfFile] if the end of the channel is reached before completing the match,
+ *         [SkipResult.Invalid] if the sequence does not match the channel's bytes.
+ */
+public suspend fun ByteReadChannel.trySkip(byteString: ByteString): SkipResult {
```

Comment:
Would the not skipped bytes be returned to the channel (if the result is `SkipResult.Invalid`)?

Reply:
Actually, I ended up using `skipIfFound` everywhere for this reason, which only consumes the bytes if the string is discovered.  I'll just remove the unused skip functions.